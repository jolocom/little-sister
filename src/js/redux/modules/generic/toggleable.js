import * as _ from 'lodash'
import { action } from '../'

export function genericReducer(actions, {initialState}) {
  return function reducer(state = initialState, action = {}) {
    switch (action.type) {
      case actions.show.id:
        return true
      case actions.hide.id:
        return false
      case actions.toggle.id:
        return !state
      default:
        return state
    }
  }
}

export default function creator(prefix, type, {initialValue = false}) {
  const camelCaseType = _.camelCase(type)
  const actions = {
    show: action(prefix, 'show' + camelCaseType, {
      expectedParams: []
    }),
    hide: action(prefix, 'hide' + camelCaseType, {
      expectedParams: []
    }),
    toggle: action(prefix, 'toggle' + camelCaseType, {
      expectedParams: []
    })
  }
  return {
    reducer: genericReducer(actions, {initialState: initialValue}),
    actions
  }
}
