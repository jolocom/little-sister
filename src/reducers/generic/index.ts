import { AnyAction } from 'redux'
import { pickBy } from 'ramda'

export const APPWRAP_UPDATE_CONFIG = 'APPWRAP_UPDATE_CONFIG'
export const APPWRAP_SHOW_LOADER = 'APPWRAP_SHOW_LOADER'
export const APPWRAP_REGISTER_CONFIG = 'APPWRAP_REGISTER_CONFIG'
export const APPWRAP_UNREGISTER_CONFIG = 'APPWRAP_UNREGISTER_CONFIG'
export const APPWRAP_SET_LOCKED = 'APPWRAP_SET_LOCKED'
export const APPWRAP_SET_DISABLE_LOCK = 'APPWRAP_SET_DISABLE_LOCK'

export interface AppWrapConfig {
  readonly withoutStatusBar: boolean
  readonly loading: boolean
  readonly dark: boolean
  readonly secondaryDark: boolean
}
const initialAppWrapAttrs: AppWrapConfig = {
  loading: false,
  withoutStatusBar: false,
  dark: false,
  secondaryDark: false,
}

export const isAppWrapConfigAttr = (key: string) => initialAppWrapAttrs[key] !== undefined
export const pickAppWrapConfigAttrs: (o: any) => AppWrapConfig =
  pickBy((v, k) => isAppWrapConfigAttr(k))

export interface AppWrapState {
  readonly locked: boolean
  readonly disableLock: boolean
  readonly appWrapConfig: AppWrapConfig
  readonly appWrapConfigsSet: AppWrapConfig[]
}
const initialAppWrapState: AppWrapState = {
  locked: false,
  disableLock: false,
  appWrapConfig: initialAppWrapAttrs,
  appWrapConfigsSet: []
}

// TODO(mnzaki): this code is unnecessarily complicated
// the complicated "ConfigsSet" keeping is to support nested <Wrapper>
// components
// - but we don't really need to support them
// - see 20APR[RN61.5] on notion.so:spaces/mnzaki

export const appWrapReducer = (
  state = initialAppWrapState,
  action: AnyAction,
): AppWrapState => {
  let appWrapConfig, appWrapConfigsSet
  switch (action.type) {
    case APPWRAP_REGISTER_CONFIG:
      const idx = state.appWrapConfigsSet.indexOf(action.value)
      if (idx > -1 || Object.keys(action.value).length === 0) return state
      appWrapConfig = {...state.appWrapConfig, ...action.value }
      appWrapConfigsSet = [...state.appWrapConfigsSet, action.value]
      return { ...state, appWrapConfig, appWrapConfigsSet }
    case APPWRAP_UNREGISTER_CONFIG:
      const idx2 = state.appWrapConfigsSet.indexOf(action.value)
      if (idx2 < 0) return state
      appWrapConfigsSet = state.appWrapConfigsSet.filter(s => s !== action.value)
      appWrapConfig = appWrapConfigsSet.reduce((prev, cur) => ({...prev, ...cur}), initialAppWrapAttrs)
      return { ...state, appWrapConfig, appWrapConfigsSet }

    case APPWRAP_SHOW_LOADER:
      return {...state, appWrapConfig: {...state.appWrapConfig, loading: action.value} }
    case APPWRAP_SET_LOCKED:
      return { ...state, locked: action.value }

    case APPWRAP_SET_DISABLE_LOCK:
      return { ...state, disableLock: action.value }
    default:
      return state
  }
}
