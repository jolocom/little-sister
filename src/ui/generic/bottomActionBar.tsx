import { StyleSheet, View, TouchableOpacity } from 'react-native'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import * as React from 'react'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'

const NAVIGATION_HEIGHT = 82
const NAVIGATION_CONTENT_HEIGHT = NAVIGATION_HEIGHT - 32

interface ActionBarProps {
  openScanner: () => void
}

const styles = StyleSheet.create({
  navigationContent: {
    height: NAVIGATION_CONTENT_HEIGHT,
    backgroundColor: '#fafafa',
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  navigationContentItem: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  contentLeft: {
    marginRight: 36,
  },
  contentRight: {
    marginLeft: 36,
  },
  qrCodeButton: {
    position: 'absolute',
    bottom: 6,
    height: 72,
    width: 72,
    borderRadius: 35,
    backgroundColor: JolocomTheme.primaryColorPurple,
    alignItems: 'center',
    justifyContent: 'center',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    elevation: 8,
  },
})

export class BottomActionBar extends React.Component<ActionBarProps, {}> {
  render() {
    return (
      <View display={'flex'} alignItems={'center'}>
        <View style={styles.navigationContent}>
          <View style={[styles.navigationContentItem, styles.contentLeft]} />
          <View style={[styles.navigationContentItem, styles.contentRight]} />
        </View>

        <TouchableOpacity
          style={styles.qrCodeButton}
          onPress={this.props.openScanner}
        >
          <Icon size={30} name="qrcode-scan" color="white" />
        </TouchableOpacity>
      </View>
    )
  }
}
