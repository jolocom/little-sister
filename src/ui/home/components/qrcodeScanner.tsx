import React from 'react'
import { Text, StyleSheet, Platform, BackHandler } from 'react-native'
import { Container } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { Button } from 'react-native-material-ui'

const QRScanner = require('react-native-qrcode-scanner').default



interface Props {
  onScannerSuccess: (e : any) => void
  onScannerCancel: () => void
}

interface State {
  listener: any
}

const styles = StyleSheet.create({
  buttonText: {
    color: JolocomTheme.palette.primaryColor
  }
})

export class QRcodeScanner extends React.Component<Props, State> {
  state = {
    listener: null
  }

  componentDidMount() {
    if (Platform.OS == "android" && this.state.listener == null) {
      this.setState({
        listener: BackHandler.addEventListener('hardwareBackPress', () => {
          this.props.onScannerCancel()
        })
      })
    }
  }

  componentWillUnmount() {
    this.setState({
      listener: null
    })
  }

  render() {
    const { onScannerSuccess, onScannerCancel } = this.props
    return (
      <Container>
        <QRScanner
          onRead={(e : any) => onScannerSuccess(e) }
          topContent={
            <Text>You can scan the qr code now!</Text>
          }
          bottomContent={
              <Button
                onPress={ () => onScannerCancel() }
                style={{ text: styles.buttonText }}
                text="Cancel"
              />
          }
        />
      </Container>
    )
  }
}
