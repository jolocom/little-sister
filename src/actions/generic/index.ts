import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { ThunkAction } from 'src/store'
import { AppError, ErrorCode } from 'src/lib/errors'
import settingKeys from '../../ui/settings/settingKeys'
import { SETTINGS } from '../../reducers/settings'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    await backendMiddleware.initStorage()
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()
    // locale setup
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale
    SplashScreen.hide()
    return dispatch(loadSettings(storedSettings))
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  }
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: SETTINGS.LOAD_SETTINGS,
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: SETTINGS.SET_LOCALE, value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}
