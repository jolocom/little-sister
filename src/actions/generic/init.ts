import { Linking } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import { AppError, ErrorCode } from '../../lib/errors'
import I18n from '../../locales/i18n'
import { routeList } from '../../routeList'
import { navigationActions, accountActions, genericActions } from '../../actions'
import { ThunkAction } from '../../store'
import { withLoading, withErrorScreen } from '../modifiers'
import { showErrorScreen } from '.'

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    const storedSettings = await backendMiddleware.storage.get.settingsObject()

    // locale setup
    if (storedSettings.locale) I18n.locale = storedSettings.locale.selected
    else storedSettings.locale = I18n.locale

    storedSettings.locale = I18n.locale

    await dispatch(loadSettings(storedSettings))

    const ret = await dispatch(accountActions.checkIdentityExists)

    // FIXME what happens if no identity and this is a deeplink?
    // navigationActions.handleDeepLink throws NoWallet
    // need to improve this UX

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    const deepLinkUrl = await Linking.getInitialURL()
    if (deepLinkUrl) {
      await handleDeepLink(deepLinkUrl)
    }

    await dispatch(genericActions.lockApp())
    return ret
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  } finally {
    SplashScreen.hide()
  }
}

const loadSettings = (settings: { [key: string]: any }) => ({
  type: 'LOAD_SETTINGS',
  value: settings,
})
