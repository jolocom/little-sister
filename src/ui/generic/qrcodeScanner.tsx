import React from 'react'
import { Text, StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import { Container } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { Button } from 'react-native-material-ui'
import { QrScanEvent } from 'src/ui/generic/qrcodeScanner'
import { ssoActions, navigationActions } from 'src/actions'
import I18n from 'src/locales/i18n'
import { LoadingSpinner } from './loadingSpinner'
const QRScanner = require('react-native-qrcode-scanner').default

export interface QrScanEvent {
  data: string
}

interface Props {
  loading: boolean
  onScannerSuccess: (e: QrScanEvent) => void
  onScannerCancel: () => void
}

interface State {}

const styles = StyleSheet.create({
  buttonText: {
    color: JolocomTheme.primaryColorBlack,
  },
})

export class QRcodeScanner extends React.Component<Props, State> {
  render() {
    const { loading, onScannerSuccess, onScannerCancel } = this.props
    return (
      <React.Fragment>
        {loading && <LoadingSpinner />}
        <Container>
          <QRScanner
            onRead={(e: QrScanEvent) => onScannerSuccess(e)}
            topContent={<Text>{I18n.t('You can scan the qr code now!')}</Text>}
            bottomContent={
              <Button
                onPress={onScannerCancel}
                style={{ text: styles.buttonText }}
                text={I18n.t('Cancel')}
              />
            }
          />
        </Container>
      </React.Fragment>
    )
  }
}

const mapStateToProps = (state: any) => ({
  loading: state.account.loading.toJS().loading,
})

const mapDispatchToProps = (dispatch: Function) => ({
  onScannerSuccess: (e: QrScanEvent) => dispatch(ssoActions.parseJWT(e.data)),
  onScannerCancel: () => dispatch(navigationActions.goBack()),
})

export const QRScannerContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(QRcodeScanner)
