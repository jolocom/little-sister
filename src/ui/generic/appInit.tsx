import * as React from 'react'
import { connect } from 'react-redux'

import { ThunkDispatch } from '../../store'
import { Image } from 'react-native'
import { withErrorHandler } from '../../actions/modifiers'
import { Wrapper } from '../structure'
import { AppError, ErrorCode } from '../../lib/errors'
import { showErrorScreen } from '../../actions/generic'
import { initApp } from 'src/actions/generic/init'

interface Props extends ReturnType<typeof mapDispatchToProps> {}

export class AppInitContainer extends React.Component<Props> {
  constructor(props: Props) {
    super(props)
    this.props.doAppInit()
  }

  render() {
    return (
      <Wrapper dark withoutSafeArea centered>
        <Image
          source={require('src/resources/img/splashIcons/joloLogoIcon.png')}
        />
      </Wrapper>
    )
  }
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  doAppInit: () =>
    dispatch(
      withErrorHandler(
        showErrorScreen,
        (err: Error) => new AppError(ErrorCode.AppInitFailed, err),
      )(initApp),
    ),
})

export const AppInit = connect(null, mapDispatchToProps)(AppInitContainer)
