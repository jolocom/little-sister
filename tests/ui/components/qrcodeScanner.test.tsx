import React from 'react'
import { QRcodeScanner } from 'src/ui/generic/qrcodeScanner'
import { shallow } from 'enzyme'
import mockCamera from '../../__mocks__/react-native-camera'

describe('QRcodeScanner component', () => {
  jest.mock('react-native-camera', () => mockCamera)

  it('matches the snapshot with back handler', () => {
    const BackHandler = jest.fn()

    const props = {
      onScannerSuccess: () => null,
      onScannerCancel: () => null,
      listener: BackHandler,
    }

    const rendered = shallow(<QRcodeScanner {...props} />)
    expect(rendered).toMatchSnapshot()
  })
})
