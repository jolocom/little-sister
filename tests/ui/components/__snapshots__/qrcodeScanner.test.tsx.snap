// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRcodeScanner component matches the snapshot with back handler 1`] = `
<React.Fragment>
  <Container>
    <QRCodeScanner
      bottomContent={
        <Button
          accent={false}
          disabled={false}
          icon={null}
          onLongPress={null}
          onPress={[Function]}
          primary={false}
          raised={false}
          style={
            Object {
              "text": Object {
                "color": "black",
              },
            }
          }
          testID={null}
          text="Cancel"
          upperCase={true}
        />
      }
      cameraType="back"
      checkAndroid6Permissions={false}
      fadeIn={true}
      notAuthorizedView={
        <Component
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            Camera not authorized
          </Text>
        </Component>
      }
      onRead={[Function]}
      pendingAuthorizationView={
        <Component
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            ...
          </Text>
        </Component>
      }
      permissionDialogMessage="Need camera permission"
      permissionDialogTitle="Info"
      reactivate={false}
      reactivateTimeout={0}
      showMarker={false}
      topContent={
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          You can scan the qr code now!
        </Text>
      }
    />
  </Container>
</React.Fragment>
`;
